### PR TITLE
Unmapped aggs should not run pipelines if they delegate reduction

### LINKED ISSUE #33514: Aggregations: toXContent response contains duplicate fields
While debugging a watcher issue I encountered a problem that serializing a search response returned an exception saying that there were duplicate fields. Checking the JSON resulted in duplicate fields. I do not however understand why this is not an issue when writing out a search response, maybe the `XContentHelper.convertToMap` implementation is the culprit.

This happened on 6.3.2, but I also reproduced this up to master.

This is the test I was able to reproduce the issue with

```java
    public void testFieldGetsWrittenOutTwice() throws Exception {
        // you need to add an additional index in order to trigger this (or potentially a shard)
        createIndex("foo_1");

        XContentBuilder builder = jsonBuilder().startObject().startObject("properties")
            .startObject("@timestamp").field("type", "date").endObject()
            .startObject("license").startObject("properties")
            .startObject("count").field("type", "long").endObject()
            .startObject("partnumber").field("type", "text").startObject("fields").startObject("keyword")
            .field("type", "keyword").field("ignore_above", 256)
            .endObject().endObject().endObject()
            .endObject().endObject().endObject().endObject();
        assertAcked(client().admin().indices().prepareCreate("foo_2")
            .addMapping("doc", builder).get());

        XContentBuilder docBuilder = jsonBuilder().startObject()
            .startObject("license").field("partnumber", "foobar").field("count", 2).endObject()
            .field("@timestamp", "2018-07-08T08:07:00.599Z")
            .endObject();

        client().prepareIndex("foo_2", "doc").setSource(docBuilder).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();

        client().admin().indices().prepareRefresh();

        TermsAggregationBuilder groupByLicenseAgg = AggregationBuilders.terms("group_by_license_partnumber").field("license.partnumber.keyword");
        MaxBucketPipelineAggregationBuilder peakPipelineAggBuilder =
            PipelineAggregatorBuilders.maxBucket("peak", "licenses_per_day>total_licenses");
        SumAggregationBuilder sumAggBuilder = AggregationBuilders.sum("total_licenses").field("license.count");
        DateHistogramAggregationBuilder licensePerDayBuilder =
            AggregationBuilders.dateHistogram("licenses_per_day").field("@timestamp").dateHistogramInterval(DateHistogramInterval.DAY);
        licensePerDayBuilder.subAggregation(sumAggBuilder);
        groupByLicenseAgg.subAggregation(licensePerDayBuilder);
        groupByLicenseAgg.subAggregation(peakPipelineAggBuilder);

        SearchResponse response = client().prepareSearch("foo_*").setSize(0).addAggregation(groupByLicenseAgg).get();
        // you can see in the logging output that there are actually two peak fields
        // I do not yet understand why this is not a problem in our rest layer but only in this case
        logger.info(Strings.toString(response));
        // conversion to a map fails because there are two `peak` fields
        BytesReference bytes = XContentHelper.toXContent(response, XContentType.JSON, false);
        XContentHelper.convertToMap(bytes, false, XContentType.JSON);
    }
```

### PR DESCRIPTION
**Pull request by polyfractal**
`UnmappedTerms`/`UnmappedSampler`/`UnmappedSigTerms` aggs try to delegate reduction to a sibling object that is not unmapped.  That delegated agg will run the reductions, and also reduce any pipeline aggs.  But because delegation comes before running pipelines, the `Unmapped*` agg _also_ tries to run pipeline aggs.

This causes the pipeline to run twice, and potentially double it's output in buckets which can create invalid JSON and break when converting to maps.

This fixes the issue by toggling a flag in UnmappedTerms if it delegated away reduction so that it knows not to run pipeline aggs either.

Closes #33514

### CODE REVIEW

**Review by colings86 — APPROVED** [review]
I'm not a huge fan of the fact that unmapped aggs need to do this delegation but this LGTM given thats how it works. I wonder if we can improve this to avoid having to delegate like this down the road?

**Comment by colings86** [comment]
> Another easy fix would be sorting the aggs before reduction, so that we preferentially choose mapped aggs as the first reduction object.

What would happen if all the aggs were unmapped, would we end up having to do the same delegation we do now?

> We'd probably need the unmapped aggs to implement a simple interface so we could tell them apart from other aggs,

This could also be done by having a method on InternalAggregation that is something like `isMapped()` which an unmapped agg could return false for?

What would be great is if we could find a way to make it so Unmapped aggs are not special at all and so we don't have to avoid using them for reduce or being careful about double execution

**Comment by polyfractal** [comment]
> What would happen if all the aggs were unmapped, would we end up having to do the same delegation we do now?

Same delegation process, but no need for the flag to track state.  If there are only unmapped aggs, the first one will get chosen to run.  Then it will look through all the aggs, see they are also all unmapped and then just return itself skipping reduction entirely.  Then it'll also run pipelines.

> This could also be done by having a method on InternalAggregation that is something like isMapped() which an unmapped agg could return false for?

++ that'd work too, and be cleaner.

**Comment by polyfractal** [comment]
@colings86 Ok, reworked to use a sorting approach.  Has an overall smaller change to the codebase, at the small cost of a sort.   I also removed the delegation ability from unmapped aggs, since it's not needed anymore.  If they get called to lead the reduction they will just return themselves, because there are no other mapped aggs.

There is a bit of a danger here that the caller won't know to sort first, and won't know that unmapped aggs just return themselves. Right now there's only one place that calls this code so it's reasonably contained.  But we could make the unmapped aggs throw a `UnsupportedOperationException` to avoid accidents.

Thoughts?

**Comment by polyfractal** [comment]
Jenkins, run gradle build tests

**Comment by polyfractal** [comment]
Jenkins, run gradle build tests

**Comment by polyfractal** [comment]
Huzzah, tests passing again.  Think this is ready for another look.

File: server/src/test/java/org/elasticsearch/search/aggregations/pipeline/MaxBucketIT.java [hunk 1]
```diff
@@ -475,4 +486,56 @@ public void testNested() throws Exception {
...
+     *
+     * This bug manifests as the max_bucket agg ("peak") being added to the response twice, because
+     * the pipeline agg is run twice.  This makes invalid JSON and breaks conversion to maps.
+     * The bug was caused by an UnmappedTerms being the chosen as the first reduction target.  UnmappedTerms
+     * delegate reduction to the first non-unmapped agg, which would reduce and run pipeline aggs.  But then
+     * execution returns to the UnmappedTerms and _it_ runs pipelines as well, doubling up on the values.
+     *
+     * Applies to any pipeline agg, not just max.
+     */
+    public void testFieldGetsWrittenOutTwice() throws Exception {
```
Comment (colings86) [inline server/src/test/java/org/elasticsearch/search/aggregations/pipeline/MaxBucketIT.java]: Isn't this testing that the field does NOT get written twice?

**Review by colings86 — APPROVED** [review]
Left a minor comment but this LGTM. Thanks for spending the time to work to a better solution

File: server/src/test/java/org/elasticsearch/search/aggregations/pipeline/MaxBucketIT.java (hunk 1 above)
Comment (polyfractal) [reply server/src/test/java/org/elasticsearch/search/aggregations/pipeline/MaxBucketIT.java]: Ah, yes, the name should be adjusted.  Named it while it was failing, but that doesn't make sense now that it's fixed :)

**Comment by polyfractal** [comment]
Jenkins, run gradle build tests

**Pull request merged**